README.md
# caluma-interval

[![Build Status](https://github.com/projectcaluma/caluma-interval/workflows/Tests/badge.svg)](https://github.com/projectcaluma/caluma-interval/actions?query=workflow%3ATests)
[![Coverage](https://img.shields.io/badge/coverage-100%25-brightgreen.svg)](https://github.com/projectcaluma/caluma-interval/blob/master/.coveragerc#L5)
[![Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/projectcaluma/caluma-interval)
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)

[Caluma](https://caluma.io/) companion app for periodic usage of forms.


## Periodic usage of forms

There are forms you might want to use periodically. This project adds support for this
by utilizing the `meta`-field of forms.

## Example use case

Your flux capacitor needs new plutonium once every two weeks. So you have a form
`refill-plutonium-in-flux-capacitor` you want to use every two weeks.

This can be achieved with following interval inside the meta field of this form:

```json
{
    "interval": {
        "interval": "P2W",
        "workflow_slug": "delorean-workflow"
    }
}
```

This will make sure, that exactly two weeks after the last refill, a new case will be
opened for this form.

Now let's say, it's important to you, that this case is always opened on a monday
(shortening the interval if needed). Further you want to set a start date, to make sure
no case will be opened before that.

This can be achieved with following interval:

```json
{
    "interval": {
        "interval": "2019-03-18/P2W",
        "weekday": 0,
        "workflow_slug": "delorean-workflow"
    }
}
```

## Features

 * Handle periodic usage of forms
 * Optionally set a start date
 * Optionally force a specific weekday
 * Will never start multiple cases for the same form


## Meta field

The `meta`-field is a JSONField and can be found on a variety of Caluma-objects. We're
only interested in forms though.

## Interval definition

caluma-interval checks the `meta`-field of forms for an `interval`-key.

Example:

```json
{
    "interval": {
        "interval": "2018-03-01/P1Y2M10D",
        "weekday": 1,
        "workflow_slug": "my-test-workflow"
    }
}
```

### Fields

#### interval

[ISO8601 time interval notation](https://en.wikipedia.org/wiki/ISO_8601#Time_intervals).

We use `{{startdate}}/{{duration}}`, where **startdate** is optional. If omitted, the
case will be opened immediately.

For duration we use a subset of
[ISO8601 duration](https://en.wikipedia.org/wiki/ISO_8601#Durations) without time.

You can find the regex we use below in [ValidationClass](#validationclass).

#### weekday

An optional weekday (zero-indexed integer). This makes sure, that only on this weekday
a case for this form will be started.

This will never exceed the configured interval.


#### workflow_slug

The slug for the corresponding workflow.


## Configuration
You can configure caluma_interval with environment variables or CLI arguments, whereas
CLI arguments take precedence.

```
usage: __main__.py [-h] [-c STRING] [-i STRING] [-s STRING] [-u STRING] [-d]
                   [-v]

Caluma companion app for periodic usage of forms

optional arguments:
  -h, --help            show this help message and exit
  -c STRING, --caluma-endpoint STRING
                        defaults to "http://caluma:8000/graphql"
  -i STRING, --oidc-client-id STRING
  -s STRING, --oidc-client-secret STRING
  -u STRING, --oidc-token-uri STRING
  -d, --debug           print debug messages
  -v, --version         show program's version number and exit
```

The corresponding environment variables are:
 * CALUMA_ENDPOINT - defaults to `http://caluma:8000/graphql`
 * OIDC_CLIENT_ID
 * OIDC_CLIENT_SECRET
 * OIDC_TOKEN_URI

Additional environment variables for when using the docker image:
 * CALUMA_HOST - defaults to `caluma`
 * CALUMA_PORT - defaults to `8000`

To enable oidc authentication over plain http (only for testing!), set
`OAUTHLIB_INSECURE_TRANSPORT` to `1`.


## Authentication
As Caluma uses OpenID Connect (OIDC), we need to fetch a token from the OIDC provider.

For this you have to provide a client id, client secret and a token uri.

If none of them are provided, the client will make unauthenticated requests, which
(hopefully) will fail in your prod setup.

## ValidationClass

You may want to add a custom ValidationClass to Caluma in order to validate the content
of `meta['interval']`:

```python
import re
from datetime import datetime

from caluma.core.validations import BaseValidation, validation_for
from caluma.form.schema import SaveForm
from caluma.workflow.models import Workflow
from rest_framework import exceptions


# Regex to parse ISO8601 periods WITHOUT time information
ISO8601_PERIOD_DATE_REGEX = re.compile(
    r"^P(?!$)"
    r"(\d+(?:[,\.]\d+)?Y)?"
    r"(\d+(?:[,\.]\d+)?M)?"
    r"(\d+(?:[,\.]\d+)?W)?"
    r"(\d+(?:[,\.]\d+)?D)?$"
)


def validate_interval(interval):
    interval_list = interval["interval"].split("/")
    if len(interval_list) == 2:
        try:
            datetime.strptime(interval_list[0], "%Y-%m-%d")
        except ValueError:
            raise exceptions.ValidationException("Failed to parse startdate!")
    elif len(interval_list) > 2:
        raise exceptions.ValidationException("Failed to parse interval!")

    if not ISO8601_PERIOD_DATE_REGEX.match(interval_list[-1]):
        raise exceptions.ValidationException("Failed to parse period!")


class FormIntervalValidation(BaseValidation):
    @validation_for(SaveForm)
    def validate_save_form(self, mutation, data, info):
        if "meta" not in data and "interval" not in data["meta"]:
            return data
        interval = data["meta"]["interval"]

        if "interval" not in interval:
            raise exceptions.ValidationException("Interval must be set!")

        if "workflow_slug" not in interval:
            raise exceptions.ValidationException("workflow_slug must be set!")

        try:
            Workflow.obj.get(slug=interval["workflow_slug"])
        except Workflow.DoesNotExist:
            raise exceptions.ValidationException(
                "Failed to get workflow with provided workflow_slug!"
            )

        if "weekday" in interval:
            if not interval["weekday"] >= 0 <= 6:
                raise exceptions.ValidationException(
                    "Weekday must be an integer from 0 to 6!"
                )

        validate_interval(interval['interval'])
```

## docker-compose integration

```
interval:
  image: ghcr.io/projectcaluma/caluma-interval:latest
  build:
    context: .
  depends_on:
    - caluma
```

## Contribution
We use following tools in order to standardize development and releases:

 * flake8
 * black
 * isort
 * commitlint

### pre-commit hooks

Pre commit hooks is an additional option instead of executing checks in your editor of choice.

First create a virtualenv with the tool of your choice before running below commands:

```bash
pip install pre-commit
pip install -r requirements-dev.txt -U
pre-commit install
pre-commit install --hook=pre-commit
pre-commit install --hook=commit-msg
```


CHANGELOG.md
# v3.0.0 (2022-10-26)

### Breaking

* Due to the usage of the newer style filter syntax for queries (6e63070064e2f72ae3226a919bbecf876e31d26f) Caluma V3 or newer is required

### Fix

* replace outdated filtering syntax for queries (6e63070064e2f72ae3226a919bbecf876e31d26f)

# v2.0.0 (2020-02-03)

### Fix

* update for caluma 5.x (0a392e14c4b1e209c8ffe371dbb074734a4e59c0)


caluma_interval/tests/conftest.py
import sys
from datetime import datetime

import pytest

import psycopg2
from caluma_interval.client import CalumaClient
from caluma_interval.interval import IntervalManager


@pytest.fixture
def manager():
    return IntervalManager(caluma_endpoint="http://caluma:8000/graphql")


@pytest.fixture
def client():
    return CalumaClient(endpoint="http://caluma:8000/graphql")


@pytest.fixture
def auth_client():
    return CalumaClient(
        endpoint="http://caluma:8000/graphql",
        oidc_client_id="id",
        oidc_client_secret="secret",
        oidc_token_uri="uri",
    )


@pytest.fixture()
def token():
    return {
        "access_token": "6c7dfa20995640c28e0eba82c5c88271",
        "expires_in": 300,
        "token_type": "bearer",
        "scope": ["caluma"],
        "expires_at": 1553154201.6981535,
        "expires_at_dt": datetime(2019, 3, 21, 7, 43, 21),
    }


@pytest.fixture
def create_forms(client):
    query = """\
mutation MyForm {
  firstIntSave: saveForm (input: {
    slug: "my-first-interval-form",
    name: "my first interval Form"
    meta: "{\\"interval\\": {\\"interval\\": \\"2018-03-01/P2W\\", \\"weekday\\": 1, \\"workflow_slug\\": \\"my-test-workflow\\"}}"
  }) {
    form {
      slug
    }
  }
  secondIntSave: saveForm (input: {
    slug: "my-second-interval-form",
    name: "my second interval Form"
    meta: "{\\"interval\\": {\\"interval\\": \\"2018-03-01/P2W\\", \\"weekday\\": 1, \\"workflow_slug\\": \\"my-test-workflow\\"}}"
  }) {
    form {
      slug
    }
  }
  nonIntSave: saveForm (input: {
    slug: "my-non-interval-form",
    name: "my non-interval Form"
  }) {
    form {
      slug
    }
  }
}\
"""
    return client.execute(query)


@pytest.fixture
def create_task(client):
    query = """\
mutation MySimpleTask {
  saveSimpleTask (input: {
    slug: "my-start-case-task",
    name: "my start case Task"
  }) {
    clientMutationId
  }
}\
"""
    return client.execute(query)


@pytest.fixture
def create_workflow(client):
    add_workflow = """\
mutation saveWorkFlow {
  saveWorkflow(input: {
    slug: "my-test-workflow"
    name: "my workflow"
    startTasks: ["my-start-case-task"]
    isPublished: true
    allowAllForms: true
  }) {
    clientMutationId
  }
}\
"""

    client.execute(add_workflow)


@pytest.fixture
def create_form_to_workflow(create_forms, create_task, create_workflow):
    pass


@pytest.fixture()
def cleanup_db():
    """
    Remove all rows from all touched tables in the caluma db.
    This is run after every test, to make sure we have a clean db.
    """
    yield
    conn = psycopg2.connect(
        host="db", database="caluma", user="caluma", password="caluma"
    )
    cur = conn.cursor()
    fmt_str = "DELETE FROM {};"
    tables = [
        "caluma_workflow_workitem",
        "caluma_workflow_case",
        "caluma_form_document",
        "caluma_form_form",
        "caluma_workflow_taskflow",
        "caluma_workflow_flow",
        "caluma_workflow_workflow_start_tasks",
        "caluma_workflow_workflow_allow_forms",
        "caluma_workflow_task",
        "caluma_workflow_workflow",
    ]
    for table in tables:
        delete_sql = fmt_str.format(table)
        cur.execute(delete_sql)
        conn.commit()

    cur.close()
    conn.close()


@pytest.fixture()
def sys_argv_handler():
    old_sys_argv = sys.argv
    yield
    sys.argv = old_sys_argv


caluma_interval/tests/test_cli.py
import sys

from caluma_interval.__main__ import main, parse_arguments


def test_arg_parsing():
    args = parse_arguments(["-c", "http://something", "-d"])

    assert args.caluma_endpoint == "http://something"
    assert args.debug

    args = parse_arguments([])
    assert not args.debug


def test_main(sys_argv_handler, create_form_to_workflow, cleanup_db):
    sys.argv = ["__main__.py", "-d"]
    main()

    sys.argv = ["__main__.py"]
    main()


def test_main_faulty_interval(
    client, sys_argv_handler, create_form_to_workflow, cleanup_db
):
    """
    Test handling of faulty intervals.

    We just want to know if this runs through without throwing any exception.

    This should never be needed, as Calumas ValidationClass should already
    enforce this.

    Nonetheless, here's the test:
    """
    query = """\
        mutation MyForm {
          firstIntSave: saveForm (input: {
            slug: "my-first-interval-form",
            name: "my first interval Form"
            meta: "{\\"interval\\": {\\"interval\\": \\"2018-03-01/P2Wpqli\\", \\"workflow_slug\\": \\"my-test-workflow\\"}}"
          }) {
            form {
              slug
            }
          }
        }\
    """
    client.execute(query)
    sys.argv = ["__main__.py", "-d"]
    main()


def test_main_failure(sys_argv_handler, create_form_to_workflow, cleanup_db, caplog):
    sys.argv = ["__main__.py", "-c", "http://nothing-here", "-d"]
    main()
    assert caplog.records[-1].levelname == "ERROR"


caluma_interval/tests/test_client.py
import requests
from requests_oauthlib import OAuth2Session

from caluma_interval.queries import intervalled_forms_query


def test_use_client(client):
    query = """\
query allForms {
  allForms{
    pageInfo {
      startCursor
      endCursor
    }
    edges {
      node {
        name
        meta
      }
    }
  }
}\
"""

    resp = client.execute(query)
    assert "allForms" in resp["data"]


def test_get_token(mocker, auth_client, token):
    mocker.patch.object(OAuth2Session, "fetch_token")
    OAuth2Session.fetch_token.return_value = token
    assert auth_client.get_token() == token


def test_authenticated_request(mocker, auth_client, token):
    mocker.patch.object(OAuth2Session, "fetch_token")
    OAuth2Session.fetch_token.return_value = token
    mocker.patch.object(requests, "post")

    auth_client.execute(intervalled_forms_query)
    OAuth2Session.fetch_token.assert_called()
    requests.post.assert_called_with(
        "http://caluma:8000/graphql",
        json={"query": intervalled_forms_query, "variables": "null"},
        headers={
            "Accept": "application/json",
            "Content-Type": "application/json",
            "Authorization": f"Bearer {token['access_token']}",
        },
    )


caluma_interval/client.py
import json
import logging
from datetime import datetime, timedelta

import requests
from envparse import env
from oauthlib.oauth2 import BackendApplicationClient
from requests_oauthlib import OAuth2Session

logger = logging.getLogger(__name__)


class CalumaClient:
    def __init__(
        self,
        endpoint=env("CALUMA_ENDPOINT", default="http://caluma:8000/graphql"),
        oidc_client_id=env("OIDC_CLIENT_ID", default=None),
        oidc_client_secret=env("OIDC_CLIENT_SECRET", default=None),
        oidc_token_uri=env("OIDC_TOKEN_URI", default=None),
    ):
        self.endpoint = endpoint
        self.oidc_client_id = oidc_client_id
        self.oidc_client_secret = oidc_client_secret
        self.oidc_token_uri = oidc_token_uri

        self._token = None

    def get_token(self):
        """
        If needed fetch a (new) token from the oidc provider.

        The threshold for fetching a new token is 1 minute before expiration.

        :return: dict
        """
        if not all([self.oidc_client_id, self.oidc_client_secret, self.oidc_token_uri]):
            logger.warning("Auth disabled due to missing parameters!")
            return

        thresh = datetime.now() + timedelta(minutes=1)

        if self._token is None or self._token["expires_at_dt"] <= thresh:
            client = BackendApplicationClient(client_id=self.oidc_client_id)
            oauth = OAuth2Session(client=client)
            self._token = oauth.fetch_token(
                token_url=self.oidc_token_uri,
                client_id=self.oidc_client_id,
                client_secret=self.oidc_client_secret,
            )
            self._token["expires_at_dt"] = datetime.utcfromtimestamp(
                int(self._token["expires_at"])
            )

        return self._token

    def execute(self, query, variables=None):
        return self._send(query, json.dumps(variables))

    def _send(self, query, variables):
        data = {"query": query, "variables": variables}
        headers = {"Accept": "application/json", "Content-Type": "application/json"}

        token = self.get_token()
        if token:
            headers["Authorization"] = f"Bearer {token['access_token']}"

        response = requests.post(self.endpoint, json=data, headers=headers)
        response.raise_for_status()
        return response.json()


caluma_interval/__main__.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

"""
This is the CLI for caluma_interval.
"""

import argparse
import logging
import sys

from envparse import env

from caluma_interval import client, interval

logger = logging.getLogger(__name__)
logging.basicConfig()


def parse_arguments(args):
    """
    Parse all given arguments.
    :param args: list
    :return: argparse.Namespace
    """
    parser = argparse.ArgumentParser(
        description=interval.__description__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )

    parser.add_argument(
        "-c",
        "--caluma-endpoint",
        metavar="STRING",
        type=str,
        help='defaults to "%(default)s"',
    )
    parser.set_defaults(
        caluma_endpoint=env("CALUMA_ENDPOINT", default="http://caluma:8000/graphql")
    )
    parser.add_argument("-i", "--oidc-client-id", metavar="STRING", type=str)
    parser.set_defaults(oidc_client_id=env("OIDC_CLIENT_ID", default=None))
    parser.add_argument("-s", "--oidc-client-secret", metavar="STRING", type=str)
    parser.set_defaults(oidc_client_secret=env("OIDC_CLIENT_SECRET", default=None))
    parser.add_argument("-u", "--oidc-token-uri", metavar="STRING", type=str)
    parser.set_defaults(oidc_token_uri=env("OIDC_TOKEN_URI", default=None))
    parser.add_argument(
        "-d", "--debug", action="store_true", help="print debug messages"
    )
    parser.add_argument(
        "-v", "--version", action="version", version=interval.__version__
    )

    args = parser.parse_args(args)

    return args


def main():
    """
    Main CLI function for caluma_interval.
    """
    args = parse_arguments(sys.argv[1:])

    loggers = [logger, interval.logger, client.logger]

    for l in loggers:
        if args.debug:
            l.level = logging.DEBUG
        else:
            l.level = logging.INFO

    manager = interval.IntervalManager(
        caluma_endpoint=args.caluma_endpoint,
        oidc_client_id=args.oidc_client_id,
        oidc_client_secret=args.oidc_client_secret,
        oidc_token_uri=args.oidc_token_uri,
    )
    try:
        manager.run()
    except Exception as e:
        logger.exception(e)


if __name__ == "__main__":
    main()


caluma_interval/interval.py
import logging
import re
from datetime import date, timedelta

from envparse import env
from isodate import isoerror, parse_date, parse_duration

from caluma_interval.client import CalumaClient
from caluma_interval.queries import intervalled_forms_query, start_case_mutation

from . import interval_metadata

logger = logging.getLogger(__name__)


__title__ = interval_metadata.__title__
__description__ = interval_metadata.__description__
__version__ = interval_metadata.__version__
__author__ = interval_metadata.__author__


ISO8601_PERIOD_DATE_REGEX = re.compile(
    r"""
     ^P(?!$)
     (\d+(?:[,.]\d+)?Y)?
     (\d+(?:[,.]\d+)?M)?
     (\d+(?:[,.]\d+)?W)?
     (\d+(?:[,.]\d+)?D)?
     $
 """,
    re.VERBOSE,
)


class IntervalManager:
    def __init__(
        self,
        caluma_endpoint=env("CALUMA_ENDPOINT", default="http://caluma:8000/graphql"),
        oidc_client_id=env("OIDC_CLIENT_ID", default=None),
        oidc_client_secret=env("OIDC_CLIENT_SECRET", default=None),
        oidc_token_uri=env("OIDC_TOKEN_URI", default=None),
    ):
        self.caluma_endpoint = caluma_endpoint
        self.oidc_client_id = oidc_client_id
        self.oidc_client_secret = oidc_client_secret
        self.oidc_token_uri = oidc_token_uri

        self._client = None
        self.action_count = 0

    @property
    def client(self):
        if not self._client:
            self._client = CalumaClient(
                self.caluma_endpoint,
                self.oidc_client_id,
                self.oidc_client_secret,
                self.oidc_token_uri,
            )
        return self._client

    def get_intervalled_forms(self):
        logger.debug("Fetching intervalled forms.")
        resp = self.client.execute(intervalled_forms_query)
        return resp["data"]["allForms"]["edges"]

    def start_case(self, form):
        variables = {
            "case": {
                "form": form["slug"],
                "workflow": form["meta"]["interval"]["workflow_slug"],
            }
        }
        resp = self.client.execute(start_case_mutation, variables)
        case_id = resp["data"]["startCase"]["case"]["id"]
        logger.debug(f'Started case ID "{case_id}" for form: "{form["slug"]}".')
        self.action_count += 1

    @staticmethod
    def parse_startdate(value):
        try:
            return parse_date(value)
        except isoerror.ISO8601Error:
            return False

    @staticmethod
    def parse_duration(value):
        if not re.match(ISO8601_PERIOD_DATE_REGEX, value):
            return False
        try:
            return parse_duration(value)
        # We exclude this from coverage, because it should be prevented by our
        # own regex check
        except isoerror.ISO8601Error:  # pragma: no cover
            return False

    def parse_interval(self, interval):
        def handle_parse_error(part, value):
            logger.error(f"Couldn't parse {part} {value}")
            return False, False

        interval = interval.split("/")
        if len(interval) > 2 or len(interval) < 1:
            return handle_parse_error("interval", interval)

        start = None
        if len(interval) == 2:
            start = self.parse_startdate(interval[0])
            if start is False:
                return handle_parse_error("startdate", interval[0])

        duration = self.parse_duration(interval[-1])
        if duration is False:
            return handle_parse_error("duration", interval[-1])

        return duration, start

    @staticmethod
    def get_last_run(form):
        """
        Iterate over all cases of a form and find the date of the last successful run.
        Defaults to 1900-01-01 if never run
        Returns False if a case is already running
        """
        last_run = date(1900, 1, 1)
        for document in form["documents"]["edges"]:
            case = document["node"]["case"]
            if case["status"] == "RUNNING":
                return False
            elif case["status"] == "COMPLETED":
                closed_at = parse_date(case["closedAt"])
                if closed_at > last_run:
                    last_run = closed_at
        logger.debug(f'Form: {form["slug"]} was last run on {last_run}.')
        return last_run

    @staticmethod
    def get_last_weekday(dt, weekday):
        """
        Based on a provided date, find last date on a specified weekday.
        :param dt: date
        :param weekday: int
        :return: date
        """
        gap = weekday - dt.weekday()
        if gap >= 0:
            gap -= 7
        return dt + timedelta(days=gap)

    def needs_action(self, last_run, interval, start, weekday=None):
        today = date.today()
        if start is None:
            start = today
        if start > today:
            return False
        next_run = last_run + interval
        if weekday is not None:
            if not next_run.weekday() == weekday:
                next_run = self.get_last_weekday(next_run, weekday)
        if next_run < today:
            return True

    def handle_form(self, form):
        last_run = self.get_last_run(form)
        if last_run is False:
            return
        interval, start = self.parse_interval(form["meta"]["interval"]["interval"])
        if interval is False:
            return
        weekday = form["meta"]["interval"].get("weekday")

        if self.needs_action(last_run, interval, start, weekday):
            self.start_case(form)

    def run(self):
        forms = self.get_intervalled_forms()
        logger.info(f"Got {len(forms)} intervalled forms from caluma.")
        for form in forms:
            try:
                self.handle_form(form["node"])
            except Exception:
                logger.exception(f"Couldn't process form \"{form['node']['slug']}\"!")
        logger.info(f"Started {self.action_count} case(s).")


caluma_interval/conftest.py
import sys

import pytest

import psycopg2

from .interval import IntervalManager


@pytest.fixture
def manager():
    return IntervalManager(caluma_uri="http://caluma:8000/graphql")


@pytest.fixture
def client(manager):
    return manager.client


@pytest.fixture
def create_forms(client):
    query = """\
mutation MyForm {
  firstIntSave: saveForm (input: {
    slug: "my-first-interval-form",
    name: "my first interval Form"
    meta: "{\\"interval\\": {\\"interval\\": \\"2018-03-01/P2W\\", \\"weekday\\": 1, \\"workflow_slug\\": \\"my-test-workflow\\"}}"
  }) {
    form {
      slug
    }
  }
  secondIntSave: saveForm (input: {
    slug: "my-second-interval-form",
    name: "my second interval Form"
    meta: "{\\"interval\\": {\\"interval\\": \\"2018-03-01/P2W\\", \\"weekday\\": 1, \\"workflow_slug\\": \\"my-test-workflow\\"}}"
  }) {
    form {
      slug
    }
  }
  nonIntSave: saveForm (input: {
    slug: "my-non-interval-form",
    name: "my non-interval Form"
  }) {
    form {
      slug
    }
  }
}\
"""
    return client.execute(query)


@pytest.fixture
def create_task(client):
    query = """\
mutation MySimpleTask {
  saveSimpleTask (input: {
    slug: "my-start-case-task",
    name: "my start case Task"
  }) {
    clientMutationId
  }
}\
"""
    return client.execute(query)


@pytest.fixture
def create_workflow(client):
    add_workflow = """\
mutation saveWorkFlow {
  saveWorkflow(input: {
    slug: "my-test-workflow"
    name: "my workflow"
    startTasks: ["my-start-case-task"]
    isPublished: true
    allowAllForms: true
  }) {
    clientMutationId
  }
}\
"""

    client.execute(add_workflow)


@pytest.fixture
def create_form_to_workflow(create_forms, create_task, create_workflow):
    pass


@pytest.fixture()
def cleanup_db():
    """
    Remove all rows from all touched tables in the caluma db.
    This is run after every test, to make sure we have a clean db.
    """
    yield
    conn = psycopg2.connect(
        host="db", database="caluma", user="caluma", password="caluma"
    )
    cur = conn.cursor()
    fmt_str = "DELETE FROM {};"
    tables = [
        "workflow_workitem",
        "workflow_case",
        "form_document",
        "form_form",
        "workflow_taskflow",
        "workflow_flow",
        "workflow_workflow_start_tasks",
        "workflow_workflow_allow_forms",
        "workflow_task",
        "workflow_workflow",
    ]
    for table in tables:
        delete_sql = fmt_str.format(table)
        cur.execute(delete_sql)
        conn.commit()

    cur.close()
    conn.close()


@pytest.fixture()
def sys_argv_handler():
    old_sys_argv = sys.argv
    yield
    sys.argv = old_sys_argv
